File: hw2/CsharpTasks/TransitiveClosure/DotGenerator.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace TransitiveClosure
{
    static class DotGenerator
    {
        public static void createDot(Boolean[][] origin, Boolean[][] result, String output) {
            int i, j;
            List<String> lines = new List<String>() { "digraph G {" };
            for (i = 0; i < result.Length; ++i)
                lines.Add(i.ToString() + " ;");
            for (i = 0; i < result.Length; ++i)
                for (j = 0; j < result[0].Length; ++j)
                    if (result[i][j].value)
                        lines.Add(i.ToString() + " -> " + j.ToString() + " [" + (origin[i][j].value ? "" : "color=red") + "];");
            lines.Add("}");
            File.WriteAllLines(output, lines);
        }
    }
}

File: hw2/FsharpTasks/TransitiveClosure/DotGenerator.fs
module DotGenerator

open Matrix

let generateDot origin result =
    
    let origin, result = toRowsList origin, toRowsList result

    let rec inner origin result orow rrow i j strlist =
        match orow with
        | [] ->
            match origin with
            | [] -> strlist
            | hd :: tl -> inner tl (List.tail result) hd (List.head result) (i + 1) 0 strlist
        | hd :: tl ->
            let hd_strlist = (string i) + " -> " + (string j) + " [" + (if hd then "" else "color=red") + "];"
            let next = inner origin result tl (List.tail rrow) i (j + 1)
            if List.head rrow
            then next (hd_strlist :: strlist)
            else next strlist

    let verticesCount = List.length origin

    let rec addVertices i strs =
        if i < verticesCount
        then addVertices (i + 1) (((string i) + " ;") :: strs)
        else strs

    let strs = (inner origin result [] [] -1 0 << addVertices 0) ["digraph G {"]

    (String.concat "\n" << List.rev) ("}" :: strs)